File: ruby/n_queens.rb
# frozen_string_literal: true

# See https://medium.com/@guy.argo/efficient-n-queens-solution-c58706170a3a
class NQueens
  def initialize(size)
    @size = size
    @queens = []
  end

  def northwest(file, rank)
    file + rank
  end

  def northeast(file, rank)
    file + @size - rank
  end

  def unsafe?(file, rank)
    @queens.any? do |f, r|
      r == rank ||
        northwest(f, r) == northwest(file, rank) ||
        northeast(f, r) == northeast(file, rank)
    end
  end

  def move!(file, rank)
    @queens.push([file, rank])
  end

  def unmove!(_file, _rank)
    @queens.pop
  end

  def solve(file = 0, &block)
    if file == @size
      yield self
    else
      @size.times do |rank|
        next if unsafe?(file, rank)

        move!(file, rank)
        solve(file + 1, &block)
        unmove!(file, rank)
      end
    end
  end

  def to_s
    board = Array.new(@size) { ['.'] * @size }
    @queens.each { |file, rank| board[rank][file] = 'Q' }
    board.reverse.map { |rank| rank.join('') }.join("\n")
  end

  def self.solve_and_display(size, display)
    i = 0
    new(size).solve do |queens|
      i += 1
      puts "\n#{i}:\n#{queens}" if display
    end
    puts i unless display
  end

  def self.solve_command
    solve_and_display(
      (ARGV[0] || 8).to_i,
      ARGV.size > 1 && ARGV[1] == 'display'
    )
  end
end

NQueens.solve_command if $PROGRAM_NAME == __FILE__

File: java/NQueens.java
// See https://medium.com/@guy.argo/efficient-n-queens-solution-c58706170a3a
class NQueens {
  int size;
  int[] queens;

  NQueens (int size) {
    this.size = size;
    this.queens = new int[size];
  }

  int northwest(int file, int rank) {
    return file + rank;
  }

  int northeast(int file, int rank) {
    return file + this.size - rank;
  }

  boolean unsafe(int file, int rank) {
    for (int f = 0; f < file; f++) {
      int r = queens[f];
      if (r == rank ||
          northwest(f, r) == northwest(file, rank) ||
          northeast(f, r) == northeast(file, rank))
          return true;
    }
    return false;
  }

  void move(int file, int rank) {
    this.queens[file] = rank;
  }

  void unmove(int file, int rank) {
  }

  void solve(int file, Solver solver) {
    if (file == this.size) {
      solver.solve(this);
    }
    else {
      for (int rank = 0; rank < this.size; rank++) {
        if (!unsafe(file, rank)) {
          move(file, rank);
          solve(file + 1, solver);
          unmove(file, rank);
        }
      }
    }
  }

  interface Solver {
    void solve(NQueens solution);
  }

  static class Printer implements Solver {
    public int count = 0;

    public void solve(NQueens solution) {
      System.out.println(++count + ":");
      System.out.println(solution.toString());
    }
  }

  static class Counter implements Solver {
    public int count = 0;

    public void solve(NQueens solution) {
      this.count += 1;
    }
  }

  public String toString() {
    StringBuffer sb = new StringBuffer();
    for (int file = 0; file < this.size; file++) {
      int rank = this.queens[file];
      for (int r = 0; r < rank; r++)
        sb.append('.');
      sb.append('Q');
      for (int r = rank+1; r < this.size; r++)
        sb.append('.');
      sb.append("\n");
    }
    return sb.toString();
  }

  static void solve_and_display(int size, boolean display) {
    if (display) {
      new NQueens(size).solve(0, new Printer());
    } else {
      Counter counter = new Counter();
      new NQueens(size).solve(0, counter);
      System.out.println(counter.count);
    }
  }

  public static void main(String [] args) {
    solve_and_display(
      args.length == 0 ? 8 : Integer.parseInt(args[0]),
      args.length > 1 && args[1].equals("display")
    );
  }
}
